### PR TITLE
add auto retry feature

### PR DESCRIPTION
**Pull request by bowensong**
Add a build-in auto retry mechanism for recoverable network exceptions (timeout / connection closed by remote server), make it a better transparent proxy, the user (client side) doesn't have to handle some very common network exceptions.

### CODE REVIEW

File: docs/source/clientcode.rst [hunk 1]
```diff
@@ -524,6 +524,17 @@ they will execute in
 
 See the :file:`timeout` example for more details.
 
+Also, there is a automatic retry mechanism for timeout or connection closed (by server side),
+in order to use this automatically retry::
+
+    Pyro4.config.MAX_RETRIES = 3      # attempt to retry 3 times before raise the exception
+
+You can also do this on a pre-proxy basis bt setting the max retries property on the proxy::
```
Comment (irmen) [inline docs/source/clientcode.rst]: bt->by

**Comment by irmen** [comment]
Thanks for this very thorough pull request, that's awesome.  I'll have a look at it soon!

**Comment by bowensong** [comment]
Sorry for the typo, I'll fix it soon.

**Comment by bowensong** [comment]
Got an error in python3:
```
======================================================================

ERROR: testTimeoutCall (PyroTests.test_server.ServerTestsMultiplexNoTimeout)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/build/irmen/Pyro4/tests/PyroTests/test_server.py", line 878, in testTimeoutCall

    self.assertRaises(Pyro4.errors.TimeoutError, p.delay, 1)

  File "/opt/python/3.2.5/lib/python3.2/unittest/case.py", line 557, in assertRaises

    callableObj(*args, **kwargs)

  File "../src/Pyro4/core.py", line 176, in __call__

    raise  # raise the last exception

RuntimeError: No active exception to reraise
```

But no idea why...

I added a `print` to the `except` part, like:
```
try:
    return self.__send(self.__name, args, kwargs)
except (errors.ConnectionClosedError, errors.TimeoutError):
    print('exception')
```
And then I got the print result, but python still saying "RuntimeError: No active exception to reraise", anyone can help? thanks

**Comment by irmen** [comment]
That was a weird error, I can't recall ever having seen that before!

**Comment by bowensong** [comment]
BTW, I think there's one more thing that worth to mention, because of the new parameter `_pyroMaxRetries`, this version will not able to load previous version pickled `Proxy` objects.

**Comment by irmen** [comment]
I haven't done this in the past either -- for instance when the handshake was introduced.
For simplicity I assume people are not persisting their proxies using pickle.

**Pull request merged**